platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,176 @@
+
+**Important**: It is the responsibility of the client code to ensure that the state is released once it is no longer needed.
+
+Once the reservation count reaches 0, the state becomes eligible for garbage collection.
+
+### Java classes
+
+- `com.swirlds.state.State` has the following lifecycle-related methods:
+  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
+  - `void computeHash()`: computes the hash of the immutable state.
+  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
+  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.
+
+These methods correspond to the state lifecycle as follows:
+
+[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)
+
+All of these methods should be moved do a separate class, which will be responsible for managing the state lifecycle. This design will provide clearer separation of concerns.
+`com.swirlds.state.State` should only be responsible for accessing and modifying the state, and the state lifecycle should be managed externally.
+
+The `SwirldStateManager` class requires refactoring. The current implementation has two sets of responsibilities:
+- Keeping track of the references to the latest immutable state and the latest mutable state. This includes creating a copy of the state and updating the references.
+  This functionality should reside in `swirlds-state-api`.
+
+- Handling Platform-related events (belongs to `swirlds-platform-core`):
+  - Handling consensus rounds.
+  - Sealing consensus rounds.
+  - Keeping the `freeze time` parameter up-to-date.
+  - Providing state for signing.
+  - Initializing the state by extracting it from the signed state.
+
+These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should offer all necessary mechanisms for the platform to interact with the state and its lifecycle.
+
+Therefore, `swirlds-state-api` needs a `StateLifecycleManager` interface with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`  that will have the following responsibilities:
+- Create/load an initial state as a starting point. If no snapshot is available, it should create a new state. 
+- Managing references to the latest immutable state and the latest mutable state.
+- Restrict mutability to a single state. Specifically, only one state object should be mutable at a time. To achieve this, we need to use a combination of `StateLifecycleManager.copy` method and `NonCopyableState` class.
+- Load snapshots from the disk. A state loaded from a snapshot must be immutable.
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+  /**
+   * Get the latest immutable state.
+   */
+  State getLatestImmutableState();
+
+  /**
+   * Get the mutable state.
+   */
+  State getMutableState();
```

Comment:
A question on this came up in Block Node design discussion.
If I have a `mutable` state, make a large set of changes, then decide to _discard_ those changes, can I discard the `mutable` state, and get a _new_ `mutable` state from the `last immutable` state?

Functionally (using this API), can I efficiently revert back to the last immutable state and then start making changes again?

The use case here is if I make a bunch of state changes in Block Node, but then something causes the block stream to stop and have to retry the current block (network interruption, failed block proof, etc...), I need a way go back to the starting state and begin applying the same block again.

Reply:
>If I have a mutable state, make a large set of changes, then decide to discard those changes, can I discard the mutable state, and get a new mutable state from the last immutable state?

If we use the pattern established for the `MerkleNode` instances and their `copy` methods, we can't. This is one-way street. That is, once you create a copy of a mutable instance it becomes immutable, and only the copy is mutable. The framework doesn't allow to create a copy of an immutable instance
`com.swirlds.state.merkle.MerkleStateRoot#copy`:

```
   public T copy() {
        throwIfImmutable();
        throwIfDestroyed();
        setImmutable(true);
        return copyingConstructor();
    }
```

>Functionally (using this API), can I efficiently revert back to the last immutable state and then start making changes again?

I think we can support this use case but it require a tiny bit more effort as it deviates from the normal usage of a copy method.

Reply:
We'll need that revert-to-last-immutable capability in Block Node to handle when a block fails validation or other interruptions, so we would very much appreciate if that could be included in the design for this API.

Reply:
Done

Reply:
Thank you.